src/USER-RANN/pair_rann.h

```diff
@@ -0,0 +1,174 @@
+    RANN::Activation * create_activation(const char *);
+
+    //global variables
+    int nelements;                // # of elements (distinct from LAMMPS atom types since multiple atom types can be mapped to one element)
+    int nelementsp;        // nelements+1
+    char **elements;              // names of elements
+    char **elementsp;        // names of elements with "all" appended as the last "element"
+    double *mass;                 // mass of each element
+    double cutmax;        // max radial distance for neighbor lists
+    int *map;                     // mapping from atom types to elements
+    int *fingerprintcount;    // static variable used in initialization
+    int *fingerprintlength;       // # of input neurons defined by fingerprints of each element.
+    int *fingerprintperelement;   // # of fingerprints for each element
+    bool doscreen;//screening is calculated if any defined fingerprint uses it
+    bool allscreen;//all fingerprints use screening so screened neighbors can be completely ignored
+    bool dospin;
+    int res;//Resolution of function tables for cubic interpolation.
+    int memguess;
+    double *screening_min;
+    double *screening_max;
+    bool **weightdefined;
+    bool **biasdefined;
+    int nmax1;
+    int nmax2;
+    int fmax;
+    int fnmax;
+    //memory actively written to during each compute:
+    double *xn,*yn,*zn,*Sik,*dSikx,*dSiky,*dSikz,*dSijkx,*dSijky,*dSijkz,*sx,*sy,*sz,**dSijkxc,**dSijkyc,**dSijkzc,*dfeaturesx,*dfeaturesy,*dfeaturesz,*features;
+    double *layer,*sum,*sum1,**dlayerx,**dlayery,**dlayerz,**dlayersumx,**dlayersumy,**dlayersumz;
+    double **dsx,**dsy,**dsz,**dssumx,**dssumy,**dssumz;
+    int *tn,*jl;
+    bool *Bij;
+
+    struct Simulation{
+      int *id;
+    bool forces;
+    bool spins;
+    double **x;
+    double **f;
+    double **s;
+    double box[3][3];
+    double origin[3];
+    double **features;
+    double **dfx;
+    double **dfy;
+    double **dfz;
+    double **dsx;
+    double **dsy;
+    double **dsz;
+    int *ilist,*numneigh,**firstneigh,*type,inum,gnum;
+    };
+
+    struct NNarchitecture{
+    int layers;
+    int *dimensions;//vector of length layers with entries for neurons per layer
+    double **Weights;
+    double **Biases;
+    int *activations;//unused
+    int maxlayer;//longest layer (for memory allocation)
+    };
+
+    Simulation *sims;
+    NNarchitecture *net;//array of networks, 1 for each element.
+
+   private:
+    template <typename T> static RANN::Fingerprint *fingerprint_creator(PairRANN *);
+    template <typename T> static RANN::Activation *activation_creator(PairRANN *);
+    //new functions
+    void allocate(std::vector<std::string>);//called after reading element list, but before reading the rest of the potential
+    void read_file(char *);//read potential file
+    void strip_comments(char *,FILE *fp);
+    void read_atom_types(std::vector<std::string>,std::vector<std::string>);
```

Comment:
are all of these functions really needed as class members? couldn't those just be static functions in the class implementation that take a pointer to the data they should update?

Reply:
Technically, no, they're not necessary as class members. However, static functions would require a lot of rewriting and I would argue it makes the code more convoluted and harder to read, especially since many of them call each other. I can't see that there is a strong reason to change this.